### PR TITLE
BM-598: Limit concurrent locked orders

### PR DESCRIPTION
**Pull request by austinabell**
Solves part of BM-598 in that we need multiple ways to limit the capacity of orders taken on by the broker, and limiting the amount of current locks is not a perfect solution but a reasonable mechanism that would be useful in the short term or to limit risk.

Will open PR for review once more thoroughly tested and will try to make sure no edge cases and minimimal overhead. Basically how this works is querying the database for orders that have been selected to be locked but not fulfilled, and then spawning tasks to handle pricing each. The capacity is calculated on an interval, every 10s, where the config and db will be checked, which will pick up when orders are fulfilled and more capacity available. Whenever pricing successful, it will reduce from the capacity to avoid extra tasks from being marked as locked.

May also refactor the design to do this, as this doesn't handle the case where the order is planned to be locked in the future very well. The issue with having the concurrent locks be monitored in the order monitor (which actually calls lock) is that it isn't clear when the broker needs to lock the order in by to make sure that it will complete it in time, because that is done in pricing.

### CODE REVIEW

File: crates/broker/src/order_picker.rs [hunk 1]
```diff
@@ -462,47 +503,93 @@ where
...
+
+                    _ = pricing_check_timer.tick() => {
+                        // Queue up orders that can be added to capacity.
+                        let order_size = if let Some(capacity) = capacity {
+                            capacity.saturating_sub(
+                                u32::try_from(pricing_tasks.len()).expect("tasks u32 overflow"),
+                            )
+                        } else {
+                            // If no maximum lock capacity, request a max of 10 orders at a time.
+                            10
```
Comment (nategraf) [inline crates/broker/src/order_picker.rs:549]: If capacity is much greater than 10, do we want to this only price 10 orders at a time?

File: crates/broker/src/order_picker.rs [hunk 2]
```diff
@@ -462,47 +503,93 @@ where
...
+                        };
+
+                        picker_copy
+                            .spawn_pricing_tasks(&mut pricing_tasks, order_size)
+                            .await
+                            .map_err(SupervisorErr::Recover)?;
```
Comment (nategraf) [inline crates/broker/src/order_picker.rs:555]: If `order_size` is zero, this seems to avoid kicking off any pricing tasks, but it does still query the DB.

File: crates/broker/src/order_picker.rs [hunk 3]
```diff
@@ -454,6 +455,46 @@ where
...
+        tasks: &mut JoinSet<Result<Option<U256>, (U256, PriceOrderErr)>>,
+        capacity: u32,
+    ) -> Result<()> {
+        if capacity == 0 {
+            return Ok(());
+        }
+
+        let order_res = self.db.update_orders_for_pricing(capacity).await?;
+
+        for (order_id, order) in order_res {
```
Comment (nategraf) [inline crates/broker/src/order_picker.rs:485]: It took me a minute to understand mechanically how this PR restricts the number of locks. It appears that is essentially done here since this loop will run 0 times if capacity is zero.

File: crates/broker/src/order_picker.rs (hunk 2 above)
Comment (austinabell) [reply crates/broker/src/order_picker.rs:555]: There is the `capacity == 0` check in that method. I can move it to wrap this call, if preferred?

File: crates/broker/src/order_picker.rs (hunk 3 above)
Comment (austinabell) [reply crates/broker/src/order_picker.rs:485]: yeah, tried to make it simpler, but couldn't think of ways to reduce more. Very open to suggestions, not ecstatic about this current design.

For this, yes, the high-level pattern is checking how many orders can be scheduled, and then spawning that many tasks. In this case though, if capacity is zero the code block isn't hit, but how it restricts further spawning is that the capacity is based on the `JoinSet` of tasks, and unless the pricing fails will count against the capacity, reducing the tasks spawned.

File: crates/broker/src/order_picker.rs (hunk 1 above)
Comment (austinabell) [reply crates/broker/src/order_picker.rs:554]: Good point!

**Review by nategraf — APPROVED** [review]
